common/src/primitives/per_thousand.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crypto::random::Rng;
+use serialization::{Decode, Encode, Error, Input};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Encode, Debug)]
+pub struct PerThousand(u16);
+
+impl PerThousand {
+    pub fn new(value: u16) -> Option<Self> {
+        if value <= 1000 {
+            Some(Self(value))
+        } else {
+            None
+        }
+    }
+
+    pub fn new_from_rng(rng: &mut impl Rng) -> Self {
+        Self(rng.gen_range(0..=1000))
+    }
+
+    pub fn value(&self) -> u16 {
+        self.0
+    }
+}
+
+impl Decode for PerThousand {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+        Self::new(u16::decode(input)?).ok_or(serialization::Error::from(
+            "PerThousand deserialization failed",
```

Comment:
Since you're decoding this to u16 first, can you add the invalid value in the error?

Reply:
It seems that I can't since `serialization::Error` has a description of type `&'static str`

Reply:
I see. Never mind.